File: elixir/triangle/triangle.exs
defmodule Triangle do
  @type kind :: :equilateral | :isosceles | :scalene

  @doc """
  Return the kind of triangle of a triangle with 'a', 'b' and 'c' as lengths.
  """
  @spec kind(number, number, number) :: { :ok, kind } | { :error, String.t }
  def kind(a, b, c) do
    [x, y, z] = Enum.sort([a, b, c])
    cond do
      Enum.any?([x, y, z], &(&1 <= 0)) -> {:error, "all side lengths must be positive"}
      x + y <= z                       -> {:error, "side lengths violate triangle inequality"}
      x == y and x == z                -> {:ok, :equilateral}
      x == y or x == z or y == z       -> {:ok, :isosceles}
      true                             -> {:ok, :scalene}
    end
  end
end

File: typescript/triangle/triangle.ts
export default class Triangle {
  sides: number[];

  constructor(...sides: number[]) {
    this.sides = sides.sort((leftSide, rightSide) => leftSide - rightSide);
  }

  kind() {
    if (this.isIllegal(this.sides)) {
      throw new Error();
    }

    if (this.isEquilateral(this.sides)) {
      return 'equilateral';
    } else if (this.isIsosoceles(this.sides)) {
      return 'isosceles';
    } else {
      return 'scalene';
    }
  }

  isEquilateral([sideOne, _sideTwo, sideThree]: number[]): boolean {
    return sideOne === sideThree;
  }

  isIsosoceles([sideOne, sideTwo, sideThree]: number[]): boolean {
    return sideOne === sideTwo || sideTwo === sideThree;
  }

  isIllegal([sideOne, sideTwo, sideThree]: number[]): boolean {
    const noSize = sideOne + sideTwo + sideThree === 0;

    const negativeSize = [sideOne, sideTwo, sideOne].some(
      side => Math.sign(side) === -1
    );

    const inEqual = !(sideOne + sideTwo >= sideThree);

    return noSize || negativeSize || inEqual;
  }
}
